<div class="alert alert-block" style="border: 1px solid #455A64;background-color:#ECEFF1;padding:5px;font-size:0.9em;">
본 자료와 관련 영상 컨텐츠는 저작권법 제25조 2항에 의해 보호를 받습니다. <br>본 컨텐츠 및 컨텐츠 일부 문구 등을 외부에 공개하거나, 요약해서 게시하지 말아주세요.<br>
    <a href='https://www.fun-coding.org'>잔재미코딩(www.fun-coding.org) Dave Lee</a>
</div>

<div class="alert alert-block" style="border: 2px solid #E65100;background-color:#FFF3E0;padding:10px">
<font size="4em" style="font-weight:bold;color:#BF360C;">코딩 테스트 학습 팁</font><br><br>
<font size="3em" style="color:#BF360C;">알고리즘을 익힌 후, 해당 알고리즘과 관련된 문제만 연이어서 쭈욱 풀어보는 방식이 가장 좋습니다</font><br>
<font size="3em" style="color:#BF360C;">처음 알고리즘을 익힐 때는 알고리즘을 익힐 때 풀이한 문제를 중심으로 하되,</font><br>
<font size="3em" style="color:#BF360C;">필요하면 감을 가지기 위해 여기에 최대한 가장 쉬운 한 두 문제를 푸시고,</font>
<font size="3em" style="color:#BF360C;">전체 알고리즘을 다 익힌 후, 가능하면</font><br> 
<font size="3em" style="color:#BF360C;">각 알고리즘별로 며칠동안 해당 알고리즘 관련 문제만 묶어서 연이어 풀어보세요!</font>
</div>

### 실전 코딩 테스트 - 동적 계획법

### 0. 세상에서 가장 쉽게! 코딩 패턴 익히기 (동적 계획법)
- 할 수 있따!!!!
```python
N = 7 # 출력해야 하는 N 번째
dp = [0] * (최대치 + 1)
dp[1] = 1 # 1번째의 결과값
dp[2] = 2 # 2번째의 결과값
for index in range(3, 최대치 + 1): # (가장 처음 점화식으로 계산되는 번째, 최대치 + 1)
    dp[index] = dp[index - 1] + dp[index - 2] # 점화식
print (dp[N]) # N 번째 값 출력
```

<div class="alert alert-block" style="border: 2px solid #E65100;background-color:#FFF3E0;padding:10px">
<font size="4em" style="font-weight:bold;color:#BF360C;">연습1</font><br><br>
<font size="3em" style="color:#BF360C;">점화식: dp[index] = dp[index - 1] + dp[index - 2]</font><br>
<font size="3em" style="color:#BF360C;">초기값: dp[1] = 2,  dp[2] = 4</font><br>
<font size="3em" style="color:#BF360C;">N 이 7번째 값 출력하기</font>    
</div>

In [2]:
N = 7
dp = [0] * 8
dp[1] = 1
dp[2] = 4
for index in range(3, 8):
    dp[index] = dp[index - 1] + dp[index - 2]
print(dp[N])

37


<div class="alert alert-block" style="border: 2px solid #E65100;background-color:#FFF3E0;padding:10px">
<font size="4em" style="font-weight:bold;color:#BF360C;">연습2</font><br><br>
<font size="3em" style="color:#BF360C;">점화식: dp[index] = dp[index - 1] + dp[index - 3]</font><br>
<font size="3em" style="color:#BF360C;">초기값: dp[1] = 3,  dp[2] = 5, dp[3] = 7</font><br>
<font size="3em" style="color:#BF360C;">N 이 20번째 값 출력하기</font>    
</div>

In [3]:
N = 20
dp = [0] * 21
dp[1] = 3
dp[2] = 5
dp[3] = 7
for index in range(4, 21):
    dp[index] = dp[index - 1] + dp[index - 3]
print(dp[N])

4618


<div class="alert alert-block" style="border: 2px solid #E65100;background-color:#FFF3E0;padding:10px">
<font size="4em" style="font-weight:bold;color:#BF360C;">연습3</font><br><br>
<font size="3em" style="color:#BF360C;">점화식: dp[index] = dp[index - 1] + dp[index - 2] + dp[index - 3]</font><br>
<font size="3em" style="color:#BF360C;">초기값: dp[1] = 1,  dp[2] = 2, dp[3] = 3</font><br>    
<font size="3em" style="color:#BF360C;">N 이 50번째 값 출력하기</font>    
</div>

In [4]:
N = 50
dp = [0] * 51
dp[1] = 1
dp[2] = 2
dp[3] = 3
for index in range(4, 51):
    dp[index] = dp[index - 1] + dp[index - 2] + dp[index - 3]
print(dp[N])

8864740270458


### 0. 연습문제: 피보나치수열 https://www.acmicpc.net/problem/2747

<div class="alert alert-block" style="border: 2px solid #E65100;background-color:#FFF3E0;padding:10px">
<font size="4em" style="font-weight:bold;color:#BF360C;">코딩 테스트 팁</font><br><br>
<font size="3em" style="color:#BF360C;">일반적인 동적 계획법 문제는 <b>가장 적은 경우의 수부터 계산을 해본 후, 패턴을 찾아,</b> 식(점화식)을 세우는 것이 핵심!</font><br>
</div>

In [1]:
import sys, io
file = open("2747_inputs.txt", 'w')
# 다음 데이터에 그대로 여러 데이터를 복사붙여넣기 하면 됨
data = """10
"""
file.write(data)
file.close()
input_file = open("2747_inputs.txt", "r") 
sys.stdin = io.StringIO(input_file.read())

In [2]:
import sys
input = sys.stdin.readline

n = int(input())
dp = [0] * 46
dp[0] = 0
dp[1] = 1
for index in range(2, 46):
    dp[index] = dp[index - 1] + dp[index - 2]
print(dp[n])

55


### 1. 연습 문제: 2xn 타일링 https://www.acmicpc.net/problem/11726

<div class="alert alert-block" style="border: 2px solid #E65100;background-color:#FFF3E0;padding:10px">
<font size="4em" style="font-weight:bold;color:#BF360C;">코딩 테스트 팁</font><br><br>
<font size="3em" style="color:#BF360C;">일반적인 동적 계획법 문제는 <b>가장 적은 경우의 수부터 계산을 해본 후, 패턴을 찾아,</b> 식(점화식)을 세우는 것이 핵심!</font><br>
</div>

<div class="alert alert-block" style="border: 2px solid #E65100;background-color:#FFF3E0;padding:10px">
<font size="4em" style="font-weight:bold;color:#BF360C;">코딩 테스트 팁</font><br><br>
<font size="3em" style="color:#BF360C;">1 ~ 1000 특정 숫자까지 숫자 generation 하는 방법</font><br>
</div>

In [1]:
for index in range(1, 1001):
    print(index)

1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178
179
180
181
182
183
184
185
186
187
188
189
190
191
192
193
194
195
196
197
198
199
200
201
202
203
204
205
206
207
208
209
210
211
212
213
214
215
216
217
218
219
220
221
222
223
224
225
226
227
228
229
230
231
232
233
234
235
236
237
238
239
240
241
242
243
244
245
246
247
248
249
250
251
252
253
254
255
256
257
258
259
260
261
262
263
264
265
266
267
268
269
270
271
272
273
274
275
276
277


<div class="alert alert-block" style="border: 2px solid #E65100;background-color:#FFF3E0;padding:10px">
<font size="4em" style="font-weight:bold;color:#BF360C;">코딩 테스트 팁</font><br><br>
<font size="3em" style="color:#BF360C;">특정 크기 및 특정 값으로 빈 리스트 만드는 방법</font><br>
<font size="3em" style="color:#BF360C;">list comprehension 으로 작성하는 방법이 성능이 조금 더 나을 수는 있음 (이후 별도 코드 작성 팁으로 설명드릴 예정)</font><br>
</div>

In [2]:
dp = [0] * 1001

In [3]:
dp[1000]

0

### 구현 코드

In [7]:
import sys, io
file = open("inputs.txt", 'w')
# 다음 데이터에 그대로 여러 데이터를 복사붙여넣기 하면 됨
data = """9
"""
file.write(data)
file.close()
input_file = open("inputs.txt", "r") 
sys.stdin = io.StringIO(input_file.read())

55


### 2. 연습문제: 파도반수열 https://www.acmicpc.net/problem/9461

<div class="alert alert-block" style="border: 2px solid #E65100;background-color:#FFF3E0;padding:10px">
<font size="4em" style="font-weight:bold;color:#BF360C;">코딩 테스트 팁</font><br><br>
<font size="3em" style="color:#BF360C;">최근 수학 문제는 언어영역과 유사합니다 (이 부분에 적응하는데 시간이 걸릴 수 있습니다)</font><br>
<font size="3em" style="color:#BF360C;">특히 입력이 어떻게 주어지는지, 출력이 어떻게 주어지는지는 최대한 꼼꼼히 확인하고, 그대로 작성해야 합니다</font><br>
</div>

- 1, 1, 1, 2, 2, 3, 4, 5, 7, 9, 12, ...
- 규칙 찾아보기: dp[ i + 3 ] = dp [ i ] + dp [ i + 1 ]

In [8]:
import sys, io
file = open("9461_inputs.txt", 'w')
# 다음 데이터에 그대로 여러 데이터를 복사붙여넣기 하면 됨
data = """2
6
12
"""
file.write(data)
file.close()
input_file = open("9461_inputs.txt", "r") 
sys.stdin = io.StringIO(input_file.read())

In [9]:
import sys
input = sys.stdin.readline
t = int(input())
dp = [0] * 101
dp[1] = 1
dp[2] = 1   
dp[3] = 1
for index in range(1, 98):
    dp[index + 3] = dp[index] + dp[index + 1]

for _ in range(t):
    print(dp[int(input())])

3
16


In [10]:
dp[100]

888855064897

### 3. 연습문제: 01타일 https://www.acmicpc.net/problem/1904

<div class="alert alert-block" style="border: 2px solid #E65100;background-color:#FFF3E0;padding:10px">
<font size="4em" style="font-weight:bold;color:#BF360C;">코딩 테스트 팁</font><br><br>
<font size="3em" style="color:#BF360C;">규칙을 찾는 것은 경험일 수 있습니다만, 우연히 빠르게 찾게 되면 술술 풀리고, 헤메면 당황할 수 있습니다</font><br>
<font size="3em" style="color:#BF360C;">최대한 그림을 그려보면서, 힌트를 찾아보고, 힌트가 안보이면, 다른 각도로 그림을 그려보며 힌트를 찾아야 함 (멘탈 관리 필요)</font><br>
<font size="3em" style="color:#BF360C;">제출 후, 에러나면, 메모리 에러, 시간 초과 에러가 많을 수 있으므로, 이런 부분도 고려할 필요가 있음</font><br>    
</div>

- N = 1(1) : 1
- N = 2(2) : 00 11
- N = 3(3) : 100 001 111
- N = 4(4) : 0000 1100 1001 0011 1111
- N = n : (n - 2)00 (n - 1)1
> 즉 dp[n] = dp[n - 2] + dp[n - 1]

In [190]:
import sys, io
file = open("inputs.txt", 'w')
# 다음 데이터에 그대로 여러 데이터를 복사붙여넣기 하면 됨
data = """4
"""
file.write(data)
file.close()
input_file = open("inputs.txt", "r") 
sys.stdin = io.StringIO(input_file.read())

### 4. 연습문제: 연속합 https://www.acmicpc.net/problem/1912
<div class="alert alert-block" style="border: 2px solid #E65100;background-color:#FFF3E0;padding:10px">
<font size="4em" style="font-weight:bold;color:#BF360C;">코딩 테스트 팁</font><br><br>
<font size="3em" style="color:#BF360C;">전형적인 점화식이 아닌 경우도 있으므로, 이 부분도 활용할 수 있어야 함</font><br>    
</div>

In [ ]:
import sys, io
file = open("inputs.txt", 'w')
# 다음 데이터에 그대로 여러 데이터를 복사붙여넣기 하면 됨
data = """10
10 -4 3 1 5 6 -35 12 21 -1
"""
file.write(data)
file.close()
input_file = open("inputs.txt", "r") 
sys.stdin = io.StringIO(input_file.read())

### 5. 연습문제: 퇴사 https://www.acmicpc.net/problem/14501

<div class="alert alert-block" style="border: 2px solid #E65100;background-color:#FFF3E0;padding:10px">
<font size="4em" style="color:#BF360C;">S전자 코딩테스트 기출 문제</font><br>
<font size="4em" style="color:#BF360C;">역순 동적계획법 이해</font><br>
</div>

<div class="alert alert-block" style="border: 2px solid #E65100;background-color:#FFF3E0;padding:10px">
<font size="4em" style="font-weight:bold;color:#BF360C;">jupyter notebook 과 코딩 테스트 팁</font><br><br>
<font size="3em" style="color:#BF360C;">input() 함수로 일일이 여러 데이터를 입력하기는 무리가 있음</font><br>
<font size="3em" style="color:#BF360C;">input() 함수에 여러 데이터를 넣는 방법으로 파일 입력을 sys.stdin (입력 스트림) 에 넣어서 input() 에 순차적으로 넣어줄 수 있음</font><br>
<font size="3em" style="color:#BF360C;">jupyter notebook 에서는 sys.stdin (입력 스트림) 이 별도 구현되어 있어서, 위의 동작이 안됨</font><br>
<font size="3em" style="color:#BF360C;">다음과 같이 특별한 코드를 사용해서, 바로바로 여러 데이터를 가지고 테스트할 수 있음</font><br>
<font size="3em" style="color:#BF360C;">이를 통해, jupyter notebook 이 가지는 코드 작성의 장점을 활용하여, 코딩 테스트를 빠르게 익힐 수 있음</font><br>
</div>

> 코딩 팁: max(a, b) 내장함수는 a 와 b 중 높은 값을 리턴해줌

In [192]:
import sys, io
file = open("inputs.txt", 'w')
# 다음 데이터에 그대로 여러 데이터를 복사붙여넣기 하면 됨
data = """10
5 50
4 40
3 30
2 20
1 10
1 10
2 20
3 30
4 40
5 50
"""
file.write(data)
file.close()
input_file = open("inputs.txt", "r") 
sys.stdin = io.StringIO(input_file.read())

90


### 6. 연습문제: palindrome https://www.acmicpc.net/problem/10942
<div class="alert alert-block" style="border: 2px solid #E65100;background-color:#FFF3E0;padding:10px">
<font size="4em" style="color:#BF360C;">쿠X 코딩테스트 기출 문제</font><br>
<font size="4em" style="color:#BF360C;">2차원 배열 기반 동적계획법 이해</font><br>
</div>

- 앞뒤를 뒤집어도 똑같은 문자열을 팰린드롬(palindrome) 이라고 함

In [194]:
import sys, io
file = open("inputs.txt", 'w')
# 다음 데이터에 그대로 여러 데이터를 복사붙여넣기 하면 됨
data = """7
1 2 1 3 1 2 1
4
1 3
2 5
3 3
5 7
"""
file.write(data)
file.close()
input_file = open("inputs.txt", "r") 
sys.stdin = io.StringIO(input_file.read())

### 7. 연습문제: 행렬곱 https://www.acmicpc.net/problem/11049
<div class="alert alert-block" style="border: 2px solid #E65100;background-color:#FFF3E0;padding:10px">
<font size="4em" style="color:#BF360C;">난이도: 상급 문제이지만, 유사한 문제이므로 참고로 이해하기로 함</font>
</div>

- 행렬곱: (행수 X 열수) 라고 했을 때
   - (3 X 2) * (2 X 4) = (3 X 4) -> 앞의 열수와 뒤의 행수가 동일해야 함 (행렬곱 조건), 결과는 (앞의 행수, 뒤의 열수) 가 됨
   - A의 크기가 5×3이고, B의 크기가 3×2, C의 크기가 2×6인 경우,
      - AB 는 5(A의행수) X 3(A의열수) X 2(B의 열수) => 5 X 2 
      - AB X C 는 5 X 2 X 6(C의 행수)

<div class="alert alert-block" style="border: 2px solid #E65100;background-color:#FFF3E0;padding:10px">
<font size="4em" style="font-weight:bold;color:#BF360C;">2차원 데이터 자료구조의 코딩 테스트 팁</font><br><br>
<font size="3em" style="color:#BF360C;">다음과 같이 list comprehension 으로 구현 가능 (일반적으로 list comprehension 이 성능이 좋음)</font>
</div>

In [207]:
import sys, io
file = open("inputs.txt", 'w')
# 다음 데이터에 그대로 여러 데이터를 복사붙여넣기 하면 됨
data = """3
5 3
3 2
2 6
"""
file.write(data)
file.close()
input_file = open("inputs.txt", "r") 
sys.stdin = io.StringIO(input_file.read())

In [208]:
import sys
input = sys.stdin.readline

N = int(input())
matrix = [list(map(int, input().split())) for _ in range(N)]
dp = [[0]*N for index in range(N)] 

In [209]:
print ('첫번째 행렬의 행수:', matrix[0][0], '첫번째 행렬의 열수:', matrix[0][1])

첫번째 행렬의 행수: 5 첫번째 행렬의 열수: 3


### 최종 코드

In [217]:
import sys, io
file = open("inputs.txt", 'w')
# 다음 데이터에 그대로 여러 데이터를 복사붙여넣기 하면 됨
data = """3
5 3
3 2
2 6
"""
file.write(data)
file.close()
input_file = open("inputs.txt", "r") 
sys.stdin = io.StringIO(input_file.read())

### 8. 연습문제: 파일 합치기 https://www.acmicpc.net/problem/11066
<div class="alert alert-block" style="border: 2px solid #E65100;background-color:#FFF3E0;padding:10px">
<font size="4em" style="color:#BF360C;">난이도: 상급 문제이지만, 유사한 문제이므로 참고로 이해하기로 함</font>
</div>

<div class="alert alert-block" style="border: 2px solid #E65100;background-color:#FFF3E0;padding:10px">
<font size="4em" style="font-weight:bold;color:#BF360C;">시간 초과 코딩 테스트 팁</font><br><br>
<font size="3em" style="color:#BF360C;">내장함수도 시간이 걸림</font><br>
    <font size="3em" style="color:#BF360C;">다음 코드는 sum() 을 매번 실행해줌으로써, 시간 초과 발생</font><br>
<font size="3em" style="color:#BF360C;">미리 계산 결과를 저장하고, 사용하도록 동적 계획법을 적용해줄 수 있음</font></div>

In [38]:
import sys, io
file = open("inputs.txt", 'w')
# 다음 데이터에 그대로 여러 데이터를 복사붙여넣기 하면 됨
data = """2
4
40 30 30 50
15
1 21 3 4 5 35 5 4 3 5 98 21 14 17 32
"""
file.write(data)
file.close()
input_file = open("inputs.txt", "r") 
sys.stdin = io.StringIO(input_file.read())

### 9. 연습문제: 파이프 옮기기 1 https://www.acmicpc.net/problem/17070
<div class="alert alert-block" style="border: 2px solid #E65100;background-color:#FFF3E0;padding:10px">
<font size="4em" style="color:#BF360C;">보드를 기반으로 푸는 문제가 다양하므로, 동적계획법 + 보드 형태에도 익숙해지기로 함</font>
</div>
<div class="alert alert-block" style="border: 2px solid #E65100;background-color:#FFF3E0;padding:10px">
<font size="4em" style="color:#BF360C;">S전자 코딩테스트 기출 문제</font><br>
<font size="4em" style="color:#BF360C;">동적계획법+보드 형태 이해</font><br>
</div>

In [ ]:
import sys, io
file = open("inputs.txt", 'w')
# 다음 데이터에 그대로 여러 데이터를 복사붙여넣기 하면 됨
data = """6
0 0 0 0 0 0
0 1 0 0 0 0
0 0 0 0 0 0
0 0 0 0 0 0
0 0 0 0 0 0
0 0 0 0 0 0
"""
file.write(data)
file.close()
input_file = open("inputs.txt", "r") 
sys.stdin = io.StringIO(input_file.read())

<div class="alert alert-block" style="border: 1px solid #455A64;background-color:#ECEFF1;padding:5px;font-size:0.9em;">
본 자료와 관련 영상 컨텐츠는 저작권법 제25조 2항에 의해 보호를 받습니다. <br>본 컨텐츠 및 컨텐츠 일부 문구 등을 외부에 공개하거나, 요약해서 게시하지 말아주세요.<br>
    <a href='https://www.fun-coding.org'>잔재미코딩(www.fun-coding.org) Dave Lee</a>
</div>